tests/test_bootstrap.py

```diff
@@ -0,0 +1,551 @@
+        self.assertTrue(
+            bs.get_common_dir().endswith("pythonforandroid/bootstraps/common")
+        )
+
+    def test_list_bootstraps(self):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.list_bootstraps`
+        returns the expected values, which should be: `empty", `service_only`,
+        `webview` and `sdl2`
+        """
+        # prepare bootstrap and distribution
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        expected_bootstraps = {"empty", "service_only", "webview", "sdl2"}
+        set_of_bootstraps = set(Bootstrap().list_bootstraps())
+        self.assertEqual(
+            expected_bootstraps, expected_bootstraps & set_of_bootstraps
+        )
+        self.assertEqual(len(expected_bootstraps), len(set_of_bootstraps))
+
+    def test_get_bootstraps_from_recipes(self):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.
+        get_bootstraps_from_recipes` returns the expected values
+        """
+        recipes_service = {"pyjnius", "python3"}
+        bs = Bootstrap().get_bootstrap_from_recipes(recipes_service, self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        self.assertEqual(bs.name, "service_only")
+
+        # test wrong recipes
+        wrong_recipes = {"python2", "python3", "pyjnius"}
+        bs = Bootstrap().get_bootstrap_from_recipes(wrong_recipes, self.ctx)
+        self.assertIsNone(bs)
+
+    @mock.patch("pythonforandroid.bootstrap.ensure_dir")
+    def test_prepare_dist_dir(self, mock_ensure_dir):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.prepare_dist_dir`
+        successfully calls once the method `endure_dir`
+        """
+        # prepare bootstrap and distribution
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        bs.prepare_dist_dir("fake_name")
+        mock_ensure_dir.assert_called_once_with(bs.dist_dir)
+
+    @mock.patch("pythonforandroid.bootstrap.open", create=True)
+    @mock.patch("pythonforandroid.util.chdir")
+    @mock.patch("pythonforandroid.bootstrap.shutil.copy")
+    @mock.patch("pythonforandroid.bootstrap.os.makedirs")
+    @mock.patch("pythonforandroid.bootstrap.sh.ln")
+    @mock.patch("pythonforandroid.bootstrap.listdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.mkdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.rm")
+    @mock.patch("pythonforandroid.bootstrap.isdir")
+    def test_bootstrap_prepare_build_dir(
+        self,
+        mock_isdir,
+        mock_sh_rm,
+        mock_sh_mkdir,
+        mock_listdir,
+        mock_sh_ln,
+        mock_os_makedirs,
+        mock_shutil_copy,
+        mock_chdir,
+        mock_open,
+    ):
```

Comment:
Actually most of the mocks are not being hit here because `symlink_java_src` is `False`, right?

Reply:
Yes, but I think that We actually need this mocking because we need to copy all the files of the bootstrap...so `symlink_java_src` it's an extra...unless I miss something...

Reply:
What I mean is for instance `mock_sh_rm` is never called because it's only being called if `symlink_java_src` is `True`.
You can verify this with `mock_sh_rm.called` after calling `bs.prepare_build_dir()`.
Same thing with most of the mocks in this test, for instance `mock_listdir`, `mock_sh_mkdir`, `mock_isdir` and `mock_sh_ln`. So since it's not being called, we could just not mock it and the code would look easier. That's only when `symlink_java_src` is `False` of course.